Review

arpita-biswas opened the pull request.

arpita-biswas committed:
Added solutions for assignment 2

JUnit tests not done yet. Not yet tested for all the possible corner cases.

arpita-biswas committed:
Updated solution for Q2

arpita-biswas committed:
Updated Tree.java

arpita-biswas committed:
Added JUnit Tests to Q1

arpita-biswas committed:
Added JUnit tests to Q2

Thomas-de-Rivaz commented on Tree.java:
```diff
@@ -0,0 +1,107 @@
+		int rightN = leftN + 1;
+		if(leftN >= data.length){
+			//System.out.println("Left of "+data[n]+" is not there and exceeds length");
+			root.left = null;
```
I don't think you need to set the value of left and right, they will be null by default.

Thomas-de-Rivaz commented on Tree.java:
```diff
@@ -0,0 +1,107 @@
+				Node<Integer> left = new Node<Integer>(data[leftN]);
+				//System.out.println("Left of "+data[n]+" is "+data[leftN]);
+				root.left = left;
+				createTree(left, leftN, data);
```
How deep a tree do you think you could create with recursion? If I passed in an input of 0,1,MIN_INT,2,MIN_INT,....,1000,MIN_INT,MIN_INT. Would that work? If not why?

arpita-biswas replied:
In worst case (right skewed tree), for populating n elements, an array of 2^n -1 elements is required (which is not at all memory-efficient). In this case, I feel it is better to use adjacency list as input for populating the tree. Shall I re-write the code such that it also takes an adjacency list as input? Are you pointing to any other reason for it to not work?

Thomas-de-Rivaz commented on Tree.java:
```diff
@@ -0,0 +1,107 @@
+		
+		str.append("Node: "+root.data+" ");
+		if(root.left != null){
+			str.append("Left: "+root.left.data+" ");
```
If Node itself is given a toString method, this code will get a lot cleaner. Tree shouldn't care how to print an individual node.

Thomas-de-Rivaz commented on Q1.java:
```diff
@@ -0,0 +1,132 @@
+		return str.toString();
+	}
+	
+	static boolean allAncestorHelper(Node<Integer> root, int num, StringBuilder str){
```
allAncestorHelper is a complex method. Adding a comment explaining what will be appended to str would be helpful

arpita-biswas replied:
Done

Thomas-de-Rivaz commented on Q1.java:
```diff
@@ -0,0 +1,132 @@
+	public static void main(String[] args) {
+		
+		//Tests
+		testNullTree(); //When (t == null)
```
Good test choices

arpita-biswas replied:
Thank you

Thomas-de-Rivaz commented on Q2.java:
```diff
@@ -0,0 +1,209 @@
+		testNullTree(); //When (t == null)
+		testEmptyTree(); //When (t.root == null)
+		testElementsNotPresent(); //When p, q or both not present in the tree
+		testSamePQ(); //When p and q 
```
Good Tests

arpita-biswas replied:
Thank you

Thomas-de-Rivaz commented on Q2.java:
```diff
@@ -0,0 +1,209 @@
+		}
+	}
+	
+	static Result commonAncestorHelper(Node<Integer> root, int p, int q){
```
Add a comment to commonAncestorHelper detailing what the returned Result means.

arpita-biswas replied:
Done

Thomas-de-Rivaz commented on Q2.java:
```diff
@@ -0,0 +1,209 @@
+		if(root == null){
+			return new Result(null, false);
+		}
+		//System.out.println("Node: "+root.data);
```
Remember to delete all of the System.outs you used for testing.

arpita-biswas replied:
Done

Thomas-de-Rivaz commented on Q2.java:
```diff
@@ -0,0 +1,209 @@
+		//System.out.println("Node: "+root.data);
+		
+		//Check if root contains p
+		if(root.data == p){//if root contains p, then search for q in the subtree
```
There is quite a bit of duplication with p & q swapped.

arpita-biswas replied:
I have created three cases now:
If root = p = q: 
Else if root = p or root = q: 
Else:

Thomas-de-Rivaz reviewed (approved):
What do you think could be modified in assignment 2 to make it work for finding the lowest common ancestor of more than 2 nodes?

arpita-biswas committed:
Added toString() method in Node class

arpita-biswas committed:
Updated Tree.java 

overloaded populateTree() and createTree() in order to populate the tree using HashMap<Integer, Integer>() as input

arpita-biswas committed:
Updated Q1.java to address comments

arpita-biswas committed:
Updated Q2.java to address comments
